net/netdev/netdev_carrier.c

```diff
@@ -104,6 +104,18 @@ int netdev_carrier_off(FAR struct net_driver_s *dev)
       devif_dev_event(dev, NETDEV_DOWN);
       arp_cleanup(dev);
 
+#ifdef CONFIG_NET_IPv4
+      dev->d_ipaddr  = INADDR_ANY;
+      dev->d_draddr  = INADDR_ANY;
+      dev->d_netmask = INADDR_ANY;
+#endif
+
+#ifdef CONFIG_NET_IPv6
+      net_ipv6addr_copy(&dev->d_ipv6addr, &in6addr_any);
+      net_ipv6addr_copy(&dev->d_ipv6draddr, &in6addr_any);
+      net_ipv6addr_copy(&dev->d_ipv6netmask, &in6addr_any);
+#endif
```

Comment:
I definitely think that we should not touch any IP configuration in netdev layer.
Based on the PR description "Some dhcp servers respond to a dhcp offer only when they receive a dhcp discovery message that belongs to their own network segment or address 0. therefore, clear the original ip address when disconnected from the ap or ifdown." I do not understand why we can't call `netlib_set_ipv4addr` with `INADDR_ANY` before sending dhcp discovery message?